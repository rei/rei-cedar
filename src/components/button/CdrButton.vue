<script setup>
import { useCssModule, computed, useSlots } from 'vue';

import mapClasses from '../../utils/mapClasses.js';
import { responsiveModifyClass, buildBooleanClass } from '../../utils/buildClass.js';
import propValidator from '../../utils/propValidator.js';

const slots = useSlots();
const props = defineProps({
tag: {
  type: String,
  default: 'button',
  validator: (value) => propValidator(value, ['button', 'a']),
},
type: {
  type: String,
  default: 'button',
  validator: (value) => propValidator(value, ['button', 'submit', 'reset']),
},
modifier: {
  type: String,
  default: 'primary',
  validator: (value) => propValidator(value, ['primary', 'secondary', 'sale', 'dark', 'link']),
},
size: {
  type: String,
  default: 'medium',
  validator: (value) => propValidator(
    value,
    ['small', 'medium', 'large'],
  ),
},
fullWidth: {
  type: [String, Boolean],
  default: false,
  validator: (value) => {
    if (typeof value === 'string') {
      return propValidator(
        value,
        ['@xs', '@sm', '@md', '@lg'],
        false,
      );
    }
    return typeof value === 'boolean';
  },
},
iconOnly: {
  type: Boolean,
  default: false,
},
withBackground: {
  type: Boolean,
  default: false,
},
});

const baseClass = 'cdr-button';
const style = useCssModule();

/************************  Computed properties *************************/
const buttonType = computed(() => props.tag === 'button' ? props.type : null);
const modifierClass = computed(() => `${baseClass}--${props.modifier}`);
const fullWidthClass = computed(() => !props.iconOnly && props.fullWidth
  && buildBooleanClass(baseClass, props.fullWidth, 'full-width'));
const sizeClass = computed(() => (!props.iconOnly
  ? responsiveModifyClass(baseClass, '', props.size)
  : `cdr-button--icon-only-${props.size}`));
const iconLeftClass = computed(() => slots['icon-left'] && slots.default
  && `${baseClass}--has-icon-left`);
const iconRightClass = computed(() => slots['icon-right'] && slots.default
  && `${baseClass}--has-icon-right`);
const iconOnlyClass = computed(() => props.iconOnly && `${baseClass}--icon-only`);
const withBackgroundClass = computed(() => props.iconOnly
  && props.withBackground && `${baseClass}--with-background`);

</script>
<template>
  <component
    :is="tag"
    :class="mapClasses(style,
                       baseClass,
                       modifierClass,
                       sizeClass,
                       fullWidthClass,
                       iconOnlyClass,
                       iconLeftClass,
                       iconRightClass,
                       iconOnlyClass,
                       withBackgroundClass,
    )"
    :type="buttonType"
  >
    <slot name="icon-left" />
    <slot />
    <slot name="icon-right" />
  </component>
</template>

<style lang="scss" module src="./styles/CdrButton.module.scss">
</style>
