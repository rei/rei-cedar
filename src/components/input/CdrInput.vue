<template>
  <cdr-label-standalone
    :for-id="uniqueId"
    :label="label"
    :hide-label="hideLabel"
    :required="required"
    :optional="optional"
    :disabled="disabled"
    :class="attrs.class"
  >
    <template
      #helper
      v-if="hasHelperTop"
    >
      <slot name="helper-text-top" />
    </template>
    <template
      #info
      v-if="hasInfo"
    >
      <slot name="info" />
    </template>

    <!-- default input slot -->
    <div :class="mapClasses(style, 'cdr-input-wrap', focusedClass)">
      <textarea
        v-if="rows && rows > 1"
        :rows="rows"
        :class="mapClasses(style,
                           baseClass,
                           multilineClass,
                           preIconClass,
                           postIconClass,
                           postIconsClass,
                           errorClass,
                           backgroundClass,
                           sizeClass,
        )"
        :id="uniqueId"
        :disabled="disabled"
        :aria-required="required || null"
        :aria-invalid="!!error || null"
        :aria-errormessage="(!!error && `${uniqueId}-error`) || null"
        v-bind="inputAttrs"
        :aria-describedby="describedby || null"
        @focus="isFocused = true"
        @blur="isFocused = false"
        v-model="inputModel"
      />
      <input
        v-else
        :type="type"
        :class="mapClasses(style,
                           baseClass,
                           preIconClass,
                           postIconClass,
                           postIconsClass,
                           errorClass,
                           backgroundClass,
                           sizeClass,
        )"
        :id="uniqueId"
        :disabled="disabled"
        :aria-required="required || null"
        :aria-invalid="!!error || null"
        :aria-errormessage="(!!error && `${uniqueId}-error`) || null"
        v-bind="inputAttrs"
        :aria-describedby="describedby || null"
        @focus="isFocused = true"
        @blur="isFocused = false"
        v-model="inputModel"
      >
      <span
        v-if="hasPreIcon"
        :class="style['cdr-input__pre-icon']"
      >
        <slot name="pre-icon" />
      </span>

      <span
        v-if="hasPostIcon"
        :class="style['cdr-input__post-icon']"
      >
        <slot name="post-icon" />
      </span>
    </div>


    <template
      #info-action
      v-if="hasInfoAction"
    >
      <slot name="info-action" />
    </template>

    <template #helper-text-bottom v-if="hasHelperBottom && !error">
      <span
        :id="`${id}-helper-text-bottom`"
        :class="style['cdr-input__helper-text']"
      >
        <slot name="helper-text-bottom" />
      </span>
    </template>

    <template #error v-if="error">
      <cdr-form-error
        :error="error"
        :role="errorRole"
        :id="`${uniqueId}-error`"
        v-if="error"
      >
        <template #error>
          <slot name="error" />
        </template>
      </cdr-form-error>
    </template>
  </cdr-label-standalone>
</template>

<script>
export default {
  inheritAttrs: false,
  customOptions: {}
}
</script>
<script setup>
import { useAttrs, useSlots, useCssModule, computed, ref } from 'vue';
import propValidator from '../../utils/propValidator';
import CdrLabelStandalone from '../labelStandalone/CdrLabelStandalone.vue';
import CdrFormError from '../formError/CdrFormError.vue';
import sizeProps from '../../props/size.js';
import backgroundProps from '../../props/background.js';
import mapClasses from '../../utils/mapClasses.js';
import uid from '../../utils/uid.js';

const emit = defineEmits(['update:modelValue'])
const props = defineProps({
    /**
     * `id` for the input that is mapped to the label `for` attribute.
    */
    id: {
      type: String   
    },
    /**
     *  'type' attribute for the input as defined by w3c.
     *  Only supporting text|email|number|password|search|url.
     *  The increment/decrement webkit psuedo element is hidden for number.
    */
    type: {
      type: [String],
      default: 'text',
      validator: (value) => propValidator(
        value,
        ['text', 'email', 'number', 'password', 'search', 'url', 'tel'],
      ),
    },
    /**
     * Label text. This is required for a11y even if hiding the label with `hideLabel`.
    */
    label: {
      type: String,
      required: true,
    },
    // sets default attrs for inputs that should use a numeric keyboard but are not strictly "numbers" (security code, CC number, postal code)
    numeric: {
      type: Boolean,
      default: false,
    },
    /**
     * Removes the label element but sets the input `aria-label` to `label` text for a11y.
    */
    hideLabel: Boolean,
    /**
     * Number of rows for input.  Converts component to text-area if rows greater than 1.
    */
    rows: Number,
    // Set which background type the input renders on
    background: backgroundProps,
    size: sizeProps,

    /**
     * Override the error message role, default is `status`.
     */
    errorRole: {
      type: String,
      required: false,
      default: 'status',
    },
    // Set error styling
    error: {
      type: [Boolean, String],
      default: false,
    },
    disabled: Boolean,
    required: Boolean,
    optional: Boolean,
    modelValue: {
      type: [String, Number, Function],
    },
})
const baseClass = 'cdr-input';
// TODO: delete un-used hasSlot props
const isFocused = ref(false);
const slots = useSlots();
const attrs = useAttrs();
const hasHelperTop = slots['helper-text-top'];
const hasHelperBottom = slots['helper-text-bottom'];
const hasPreIcon = slots['pre-icon'];
const hasPostIcon = slots['post-icon'];
const hasPostIcons = hasPostIcon && slots['post-icon']().length > 1;
const hasInfo = slots.info;
const hasInfoAction = slots['info-action'];

const uniqueId = props.id ? props.id : uid();
const multilineClass = computed(() => props.rows > 1 && 'cdr-input--multiline');
const preIconClass = computed(() => hasPreIcon && 'cdr-input--preicon');
// TODO: make one class for this? if possible? there must have been a reason i did it like this.....
const postIconClass = computed(() => hasPostIcon && 'cdr-input--posticon');
const postIconsClass = computed(() => hasPostIcons && 'cdr-input--posticons');
const errorClass = computed(() => props.error && 'cdr-input--error');
const backgroundClass = computed(() => `cdr-input--${props.background}`);
const sizeClass = computed(() => props.size && `${baseClass}--${props.size}`);
const focusedClass = computed(() => isFocused.value && 'cdr-input--focus');

const describedby = computed(() => {
  const helperText = [
    slots['helper-text-top'] ? `${uniqueId}-helper-text-top` : '',
    slots['helper-text-bottom'] ? `${uniqueId}-helper-text-bottom` : '',
    attrs['aria-describedby'],
  ].filter((x) => x).join(' ');

  if(props.error) {
    return `${uniqueId}-error`
  }

  return helperText;
})

const attrsWithClassExcluded = computed(()=>{
  let returnObj = {}
  for (const attr in attrs) {
    if (attr !== 'class') {
      returnObj[attr] = attrs[attr]
    }
  }
  return returnObj;
})

const inputAttrs = computed(() => {
  const isNum = props.numeric || props.type === 'number';
  return {
    id: uniqueId,
    autocorrect: 'off',
    spellcheck: 'false',
    autocapitalize: 'off',
    pattern: (isNum && '[0-9]*') || null,
    inputmode: (isNum && 'numeric') || null,
    novalidate: isNum || null,
    ...attrsWithClassExcluded.value,
  };
});
const style = useCssModule();
 const inputModel = computed({
  get() {
    return props.modelValue
  },
  set(newValue) {
    emit('update:modelValue', newValue)
  }
})
</script>
<style lang="scss" module src="./styles/CdrInput.module.scss">
</style>
