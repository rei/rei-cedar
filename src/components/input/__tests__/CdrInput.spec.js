import { mount } from '../../../../test/vue-jest-style-workaround.js';
import sinon from 'sinon';
import { h } from '@vue/runtime-core';
import CdrInput from '../CdrInput.vue';
import CdrButton from '../../button/CdrButton.vue';

const baseComponentPattern = {
  propsData: {
    label: 'Label Test',
    id: 'renders',
  }
}

describe('CdrInput', () => {
  describe('basic input component', () => {
    let wrapper;
    let input;
    beforeEach(() => {
      wrapper = mount(CdrInput, { ...baseComponentPattern });
      input = wrapper.find('input');
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders an input element', () => {
      expect(input.element.tagName).toBe('INPUT');
    });

    it('sets default input type correctly', () => {
      expect(input.attributes('type')).toBe('text');
    });

    it('does not apply aria-describedby if attr or helper slots are not present', () => {
      expect(input.attributes('aria-describedby')).toBe(undefined);
    });

    it('emits an input event', () => {
      input.setValue('foo');
      expect(wrapper.emitted()['update:modelValue'][0][0]).toBe('foo');
    });
    
    describe('with prop type of "number"', () => {
      beforeEach(() => {
        wrapper.setProps({ type: 'number' })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });


      it('has a inputmode attribute of "numeric"', () => {
        expect(input.attributes('inputmode')).toBe('numeric');
      });
    });

    describe('with error prop set', () => {
      beforeEach(() => {
        wrapper.setProps({ error: 'Something is wrong!' })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('with the "disabled" prop set to true', () => {
      beforeEach(() => {
        wrapper.setProps({ disabled: true })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets input disabled attribute correctly', () => {
        expect(input.attributes('disabled')).toBe('');
      });
    });

    describe('with the "required" prop set to true', () => {
      beforeEach(() => {
        wrapper.setProps({ required: true })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets input disabled attribute correctly', () => {
        expect(input.attributes('aria-required')).toBe('true');
      });
    });

    describe('with prop "type" set to "number"', () => {
      beforeEach(() => {
        wrapper.setProps({
          required: true,
          type: 'number',
        })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets the expected attributes', () => {
        expect(wrapper.find('input').attributes('novalidate')).toBe('');
        expect(wrapper.find('input').attributes('pattern')).toBe('[0-9]*');
        expect(wrapper.find('input').attributes('inputmode')).toBe('numeric');
        expect(wrapper.find('input').attributes('type')).toBe('number');
      });
    });

    describe('with prop "numeric" set to true', () => {
      beforeEach(() => {
        wrapper.setProps({
          required: true,
          numeric: true,
        })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets the expected attributes', () => {
        expect(wrapper.find('input').attributes('novalidate')).toBe('');
        expect(wrapper.find('input').attributes('pattern')).toBe('[0-9]*');
        expect(wrapper.find('input').attributes('inputmode')).toBe('numeric');
        expect(wrapper.find('input').attributes('type')).toBe('text');
      });
    });

    describe('with the "rows" prop set to 2', () => {
      beforeEach(() => {
        wrapper.setProps({ rows: 2 })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets multiline to textarea correctly', () => {
        expect(wrapper.find('textarea').element.tagName).toBe('TEXTAREA');
      });
    });

    describe('when "multiline" is set to true and the "rows" prop is set to 10', () => {
      beforeEach(() => {
        wrapper.setProps({
          multiLine: true,
          rows: 10
        })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets multiline rows correctly', () => {
        expect(wrapper.find('textarea').attributes('rows')).toBe('10');
      });
    });

    describe('when "type" is set to url', () => {
      beforeEach(() => {
        wrapper.setProps({ type: 'url' })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('overrides input type correctly', () => {
        expect(wrapper.find('.cdr-input').attributes('type')).toBe('url');
      });
    });

    describe('with error prop set', () => {
      beforeEach(() => {
        wrapper.setProps({ error: 'incorrect!' })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('renders text when passed as error', () => {
        expect(wrapper.find('.cdr-form-error').text()).toBe('incorrect!');
      });
    });
  });

  describe('components with attributes', () => {
    let wrapper;
    let input;
    let blurSpy;
    let pasteSpy;
    let keydownSpy;
    let focusSpy;
    beforeEach(() => {
      blurSpy = sinon.spy();
      pasteSpy = sinon.spy();
      keydownSpy = sinon.spy();
      focusSpy = sinon.spy();
      wrapper = mount(CdrInput, {
        propsData: {
          label: 'Label Test',
          id: 'test'
        },
        attrs: {
          pizza: 'time',
          name: 'testName',
          readonly: true,
          autofocus: true,
          maxlength: '20',
          placeholder: 'test placeholder',
          onBlur: blurSpy,
          onPaste: pasteSpy,
          onKeydown: keydownSpy,
          onFocus: focusSpy
        }
      });
      input = wrapper.find('.cdr-input');
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('passes all attributes to input element', () => {
      expect(input.attributes('pizza')).toBe('time');
      expect(input.attributes('name')).toBe('testName');
      expect(input.attributes('readonly')).toBe('');
      expect(input.attributes('autofocus')).toBe('');
      expect(input.attributes('maxlength')).toBe('20');
      expect(input.attributes('placeholder')).toBe('test placeholder');
    });

    it('emits a blur event', () => {
      input.trigger('blur')
      expect(blurSpy.calledOnce).toBeTruthy();
    });

    it('emits a paste event', () => {
      input.trigger('paste')
      expect(pasteSpy.calledOnce).toBeTruthy();
    });

    it('emits a keydown event', () => {
      input.trigger('keydown', { key: 'a' })
      expect(keydownSpy.called).toBeTruthy();
    });

    it('emits a focus event', () => {
      input.trigger('focus')
      expect(focusSpy.calledOnce).toBeTruthy();
    })

    it('adds focused class to wrapper on input focus and removes it on blur', async () => {
      input.trigger('focus')
      await wrapper.vm.$nextTick();
      expect(wrapper.find('.cdr-input--focus').exists()).toBeTruthy();
      input.trigger('blur')
      await wrapper.vm.$nextTick();
      expect(wrapper.find('.cdr-input--focus').exists()).toBeFalsy();
    });
  });

  describe('with "helper-text-bottom" slot', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        ...baseComponentPattern,
        slots: { 'helper-text-bottom': 'very helpful' },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders helper-text-bottom slot', () => {
      expect(wrapper.find('.cdr-input__helper-text').text()).toBe('very helpful');
    });
  });

  describe('with a pre-icon slot', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        ...baseComponentPattern,
        slots: { 'pre-icon': 'ðŸ¤ ' },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders pre-icon slot', () => {
      expect(wrapper.find('.cdr-input__pre-icon').text()).toBe('ðŸ¤ ');
    });
  });

  describe('with a post-icon slot', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        ...baseComponentPattern,
        slots: { 'post-icon': 'ðŸ˜Ž' },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders post-icon slot', () => {
      expect(wrapper.find('.cdr-input__post-icon').text()).toBe('ðŸ˜Ž');
    });

    it('adds spacing class when post-icon slot is present', () => {
      expect(wrapper.find('.cdr-input--posticon').exists()).toBe(true);
    });
  });

  describe('when multiple elements are present in post-icon slot', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        ...baseComponentPattern,
        slots: {
          'post-icon': () => [
            h(CdrButton),
            h(CdrButton)
          ],
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('adds spacing class', () => {
      expect(wrapper.find('.cdr-input--posticons').exists()).toBe(true);
    });
  });

  describe('with info action slot', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        ...baseComponentPattern,
        slots: { 'info-action': 'ðŸ¤ ' },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders info action slot', () => {
      expect(wrapper.find('.cdr-label-standalone__info-action').text()).toBe('ðŸ¤ ');
    });
  });

  describe('when error state is active', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        propsData: {
          id: 'test',
          label: 'test',
          error: true
        },
        slots: { 'error': 'whoops' },
      })
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders error slot', () => {
      expect(wrapper.find('.cdr-form-error').text()).toBe('whoops');
    });
  });

  describe(' when error state is inactive', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        propsData: {
          id: 'test',
          label: 'test',
          error: false
        },
        slots: { 'error': 'whoops' },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('does not render error slot', () => {
      expect(wrapper.find('.cdr-form-error').exists()).toBe(false);
    });
  });

  describe('when error is active and a helper-bottom-text slot has been passed', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        propsData: {
          id: 'test',
          label: 'test',
          error: true,
        },
        slots: {
          'error': 'whoops',
          'helper-text-bottom': 'not me'
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders error slot instead of bottom helper slot', () => {
      expect(wrapper.find('.cdr-form-error').text()).toBe('whoops');
      expect(wrapper.find('.cdr-input__helper-text-bottom').exists()).toBe(false);
    });
  });

  describe('with both helper-text-top and helper-text-bottom slots', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        propsData: {
          label: 'test',
          id: 'aria-test',
        },
        slots: {
          'helper-text-top': 'extremely helpful',
          'helper-text-bottom': 'very helpful',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('helper text slots are linked to input via aria-describedby', () => {
      expect(wrapper.find('input').attributes('aria-describedby')).toBe('aria-test-helper-text-top aria-test-helper-text-bottom');
    });
  });

  describe('with both helper-text-top and helper-text-bottom slots and aria-describedby set', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrInput, {
        propsData: {
          label: 'test',
          id: 'aria-test',
        },
        attrs: {
          'aria-describedby': 'foo',
        },
        slots: {
          'helper-text-top': 'extremely helpful',
          'helper-text-bottom': 'very helpful',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('dynamic aria-describedby is merged with native attr', () => {
      expect(wrapper.find('input').attributes('aria-describedby')).toBe('aria-test-helper-text-top aria-test-helper-text-bottom foo');
    });
  });
});
