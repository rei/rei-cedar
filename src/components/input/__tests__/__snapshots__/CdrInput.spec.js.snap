// Vitest Snapshot v1

exports[`CdrInput > renders correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test 
      <!--v-if-->
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    
    <div
      class="cdr-input-wrap"
    >
      <input
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--primary cdr-input--medium"
        id="renders"
        spellcheck="false"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div
    class="cdr-label-standalone__post-content"
  >
    
    
    
    
  </div>
</div>
`;

exports[`CdrInput > renders error state correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test 
      <!--v-if-->
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    
    <div
      class="cdr-input-wrap"
    >
      <input
        aria-describedby="renders-error"
        aria-errormessage="renders-error"
        aria-invalid="true"
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--error cdr-input--primary cdr-input--medium"
        id="renders"
        spellcheck="false"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div
    class="cdr-label-standalone__post-content"
  >
    
    
    
    <div
      class="cdr-form-error"
      id="renders-error"
      role="status"
    >
      <span
        class="cdr-form-error__icon"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon cdr-icon--small cdr-icon--inherit-color"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          
          
          
          <g>
            <path
              d="M15.293 16.707a1 1 0 001.414-1.414L9.121 7.707a1 1 0 10-1.414 1.414l7.586 7.586z"
              role="presentation"
            />
            <path
              clip-rule="evenodd"
              d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10zm-2 0a8 8 0 11-16 0 8 8 0 0116 0z"
              fill-rule="evenodd"
            />
          </g>
          
          <!--v-if-->
        </svg>
      </span>
      
      <span>
        Something is wrong!
      </span>
      
    </div>
    
  </div>
</div>
`;

exports[`CdrInput > renders number input correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test 
      <!--v-if-->
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    
    <div
      class="cdr-input-wrap"
    >
      <input
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--primary cdr-input--medium"
        id="renders"
        inputmode="numeric"
        novalidate=""
        pattern="[0-9]*"
        spellcheck="false"
        type="number"
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div
    class="cdr-label-standalone__post-content"
  >
    
    
    
    
  </div>
</div>
`;

exports[`CdrInput renders correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test 
      <!--v-if-->
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    
    <div
      class="cdr-input-wrap"
    >
      <input
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--primary cdr-input--medium"
        id="renders"
        spellcheck="false"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div
    class="cdr-label-standalone__post-content"
  >
    
    
    
    
  </div>
</div>
`;

exports[`CdrInput renders error state correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test 
      <!--v-if-->
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    
    <div
      class="cdr-input-wrap"
    >
      <input
        aria-errormessage="renders-error"
        aria-invalid="true"
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--error cdr-input--primary cdr-input--medium"
        id="renders"
        spellcheck="false"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div
    class="cdr-label-standalone__post-content"
  >
    
    
    
    <div
      class="cdr-form-error"
      id="renders-error"
      role="status"
    >
      <span
        class="cdr-form-error__icon"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon cdr-icon--small cdr-icon--inherit-color"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          
          
          
          <g>
            <path
              d="M15.293 16.707a1 1 0 001.414-1.414L9.121 7.707a1 1 0 10-1.414 1.414l7.586 7.586z"
              role="presentation"
            />
            <path
              clip-rule="evenodd"
              d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10zm-2 0a8 8 0 11-16 0 8 8 0 0116 0z"
              fill-rule="evenodd"
            />
          </g>
          
          <!--v-if-->
        </svg>
      </span>
      
      <span>
        Something is wrong!
      </span>
      
    </div>
    
  </div>
</div>
`;

exports[`CdrInput renders number input correctly 1`] = `
<div
  class="cdr-label-standalone"
>
  <div
    class="cdr-label-standalone__label-wrapper"
  >
    <label
      class="cdr-label-standalone__label"
      for="renders"
    >
      Label Test 
      <!--v-if-->
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="cdr-label-standalone__input-wrap cdr-label-standalone__input-spacing"
  >
    
    <div
      class="cdr-input-wrap"
    >
      <input
        autocapitalize="off"
        autocorrect="off"
        class="cdr-input cdr-input--primary cdr-input--medium"
        id="renders"
        inputmode="numeric"
        novalidate=""
        pattern="[0-9]*"
        spellcheck="false"
        type="number"
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div
    class="cdr-label-standalone__post-content"
  >
    
    
    
    
  </div>
</div>
`;
